stdlib/public/TensorFlow/Dataset.swift

```diff
@@ -41,17 +36,14 @@ public extension SingleValueDataset {
   /// - Parameter
   ///   - elements: The batch of elements.
   ///   - elementShape: The shape that the 
-  ///
-  // FIXME: Due to inability to perform shape inference during deabstraction,
-  // users must specify the element shape for today.
   @inlinable @inline(__always)
-  public init(elements: Tensor<Scalar>, elementShape: TensorShape) {
+  public init(elements: Tensor<ScalarOfElement>, elementShape: TensorShape) {
```

Comment:
nit: any reason to rename it to ScalarOfElement. Scalar seemed to be fine as is?

Reply:
Wouldn't Element be appropriate then, or is there some reason it doesn't work well?

Reply:
Oh, I see.  You've designed this to work as:  `DataSet<Float>` where the elements are implicitly `Tensor<Float>`.

Have you considered a model where users use `DataSet<Tensor<Float>>`?  This seems like it would generalize better to structured wrappers and possibly eventually aggregates of tensors.

Reply:
> Have you considered a model where users use `DataSet<Tensor<Float>>`? 

I have considered that and understand that it's the direction towards supporting arbitrarily nested tensor aggregates. That means we are going to use an unconstrained generic type `Element` eventually. But today, `#tfop` is not properly type-checked.

Let's take single value dataset for an example. If your `Element` type is `Tensor<Float>`, then the `#tfop` call site will become:
```swift
#tfop("IteratorGetNext", ...) as Element
```

When `Element` is `Tensor`, this fails tragically because `#tfop` only supports `XXXHandle` as the output type. When the output type is not `XXXHandle`, it's not type-checked but will rather rigger an assertion somewhere in TFPartition.

Today, I just wanted the `ScalarOfElement` type to be `AccelerableByTensorFlow` to make sure it works and won't assert for random types. When `#tfop` is fixed, we'll certainly move to that model.

To support arbitrary arity or aggregates, `#tfop` has to support *extracting a list of `Scalar` metatypes from an aggregate*. Because every dataset op has the following pattern:
```swift
#tfop(..., element_dtypes: [Scalar1.self, Scalar2.self, Scalar3.self, ...], element_shapes: [shape1, shape2, ...])
```

Reply:
Ok, sounds great.  Please file feature request bugs for those new #tfop capabilities :-)